pkg/ccl/changefeedccl/parquet_sink_cloudstorage.go

```diff
@@ -0,0 +1,289 @@
+	}
+
+	schema := pqexporter.NewParquetSchema(parquetColumns)
+
+	parquetWriterOptions := make([]goparquet.FileWriterOption, 0)
+
+	// TODO(cdc): We really should revisit if we should include any metadata in
+	// parquet files. There are plenty things we can include there, including crdb
+	// native column types, OIDs for those column types, etc
+	if includeParquetTestMetadata {
+		metadata, err := getMetadataForParquetFile(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+		parquetWriterOptions = append(parquetWriterOptions, goparquet.WithMetaData(metadata))
+	}
+
+	// TODO(cdc): Determine if we should parquet's builtin compressor or rely on
+	// sinks compressing. Currently using not parquets builtin compressor, relying
+	// on sinks compression
+	parquetWriterOptions = append(parquetWriterOptions, goparquet.WithSchemaDefinition(schema))
+	pqw := goparquet.NewFileWriter(buf,
+		parquetWriterOptions...,
+	)
+
+	pqww := &parquetFileWriter{
+		pqw,
+		schema,
+		parquetColumns,
+		len(parquetColumns),
+	}
+	return pqww, nil
+}
+
+func getMetadataForParquetFile(ctx context.Context, row cdcevent.Row) (map[string]string, error) {
+	metadata := make(map[string]string)
+	primaryKeyColNames := ""
+	columnNames := ""
+	if err := row.ForEachKeyColumn().Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
+		primaryKeyColNames += col.Name + ","
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	metadata["primaryKeyNames"] = primaryKeyColNames
+	if err := row.ForEachColumn().Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
+		columnNames += col.Name + ","
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	metadata["columnNames"] = columnNames
+	return metadata, nil
+}
+
+func getParquetColumnTypes(
+	ctx context.Context, row cdcevent.Row,
+) ([]pqexporter.ParquetColumn, error) {
+	typs := make([]*types.T, 0)
+	names := make([]string, 0)
+
+	if err := row.ForEachColumn().Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
+		typs = append(typs, col.Typ)
+		names = append(names, col.Name)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	parquetColumns := make([]pqexporter.ParquetColumn, len(typs)+1)
+	const nullable = true
+
+	for i := 0; i < len(typs); i++ {
+		// Make every field optional, so that all schema evolutions for a table are
+		// considered "backward compatible" by parquet. This means that the parquet
+		// type doesn't mirror the column's nullability, but it makes it much easier
+		// to work with long histories of table data afterward, especially for
+		// things like loading into analytics databases.
+		parquetCol, err := pqexporter.NewParquetColumn(typs[i], names[i], nullable)
+		if err != nil {
+			return nil, err
+		}
+		parquetColumns[i] = parquetCol
+	}
+
+	// Add the extra column which will store the type of event that generated that
+	// particular row.
+	var err error
+	parquetColumns[len(typs)], err = pqexporter.NewParquetColumn(types.Int4, parquetCrdbEventTypeColName, false)
```

Comment:
Let's not have this column as an int.
Let's have it instead as a single Byte (or string):
'u' -- for update
'c' -- for create
'd' -- for delete.